include/dr/mhp/containers/distributed_dense_matrix.hpp

```diff
@@ -0,0 +1,323 @@
+// SPDX-FileCopyrightText: Intel Corporation
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#pragma once
+
+#include "index.hpp"
+#include "segments.hpp"
+#include "rows.hpp"
+#include "subrange.hpp"
+
+namespace dr::mhp {
+
+using key_type = index<>;
+
+class distributed_matrix_partition {};
+class by_row final : public distributed_matrix_partition {};
+class by_column final : public distributed_matrix_partition {};
+class tiled final : public distributed_matrix_partition {};
+
+template <typename DM> class dm_segments : public std::span<d_segment<DM>> {
+public:
+  dm_segments() {}
+  dm_segments(DM *dm) : std::span<d_segment<DM>>(dm->segments_) { dm_ = dm; }
+
+private:
+  DM *dm_;
+}; // dm_segments
+
+template <typename T, typename Allocator> class distributed_dense_matrix {
+public:
+  using value_type = T;
+  using size_type = std::size_t;
+  using difference_type = std::ptrdiff_t;
+  using key_type = index<>;
+
+  using iterator =
+      dr::normal_distributed_iterator<dm_segments<distributed_dense_matrix>>;
+
+  distributed_dense_matrix(
+      std::size_t rows, std::size_t cols,
+      dr::halo_bounds hb = dr::halo_bounds(),
+      distributed_matrix_partition *partition = new by_row(),
+      Allocator allocator = Allocator())
+      : distributed_dense_matrix(key_type(rows, cols), hb, partition,
+                                 allocator){};
+
+  distributed_dense_matrix(
+      std::size_t rows, std::size_t cols, T fillval,
+      dr::halo_bounds hb = dr::halo_bounds(),
+      distributed_matrix_partition *partition = new by_row(),
+      Allocator allocator = Allocator())
+      : distributed_dense_matrix(key_type(rows, cols), hb, partition,
+                                 allocator) {
+
+    for (std::size_t _i = 0; _i < data_size_; _i++)
+      data_[_i] = fillval;
+  };
+
+  distributed_dense_matrix(
+      key_type shape, dr::halo_bounds hb = dr::halo_bounds(),
+      distributed_matrix_partition *partition = new by_row(),
+      Allocator allocator = Allocator())
+      : shape_(shape), segment_shape_((shape_[0] + default_comm().size() - 1) /
+                                          default_comm().size(),
+                                      shape_[1]),
+        segment_size_(std::max({segment_shape_[0] * shape_[1],
+                                hb.prev * shape_[1], hb.next * shape_[1]})),
```

Comment:
as we discussed code does not support halo being greater than segment

therefore segment_size is always `segment_shape_[0] * segment_shape_[1]`
add assertion that `segment_shape_[0] <= max(hb.prev, hb.next)` with comment that this is not yet implemented

Reply:
To be verified, whether it is or it is not implemented in halo_impl class